File: react-native/src/CampaignCache.ts
import type { MoEngageSelfHandledData } from './types';

export interface ICampaignCache {
    readonly count: number;
    readonly campaignIds: string[];
    put(campaignId: string, data: MoEngageSelfHandledData): void;
    get(campaignId: string): MoEngageSelfHandledData | undefined;
    delete(campaignId: string): void;
    clear(): void;
}

export class CampaignCache implements ICampaignCache {
    private readonly _map = new Map<string, MoEngageSelfHandledData>();

    get count(): number { return this._map.size; }
    get campaignIds(): string[] { return [...this._map.keys()]; }

    put(campaignId: string, data: MoEngageSelfHandledData): void {
        this._map.set(campaignId, data);
    }

    get(campaignId: string): MoEngageSelfHandledData | undefined {
        return this._map.get(campaignId);
    }

    delete(campaignId: string): void {
        this._map.delete(campaignId);
    }

    clear(): void {
        this._map.clear();
    }
}

File: iOS/Sources/DigiaMoEngage/cache/CampaignCache.swift
import MoEngageInApp

/// In-memory implementation of `ICampaignCache`.
///
/// Backed by a plain `Dictionary`; entries are evicted per-campaign on
/// `remove(_:)` (post-dismiss) and globally on `clear()` (teardown).
///
/// Swap this with an LRU or persistent cache by implementing `ICampaignCache`
/// and injecting it into `MoEngagePlugin` — no other code changes required.
final class CampaignCache: ICampaignCache {
    private var store: [String: InAppSelfHandledCampaign] = [:]

    func put(campaignId: String, data: InAppSelfHandledCampaign) {
        store[campaignId] = data
    }

    func get(campaignId: String) -> InAppSelfHandledCampaign? {
        store[campaignId]
    }

    func remove(campaignId: String) {
        store.removeValue(forKey: campaignId)
    }

    func clear() {
        store.removeAll()
    }

    var count: Int { store.count }

    var campaignIds: [String] { Array(store.keys) }
}
